REGRESSION (iOS 16): LocalStorage is cleared after quota error

szewai opened the pull request:
#### 5bc131992783cd2f65a8e2ed0c827eaef6df1427
<pre>
REGRESSION (iOS 16): LocalStorage is cleared after quota error
<a href="https://bugs.webkit.org/show_bug.cgi?id=245479">https://bugs.webkit.org/show_bug.cgi?id=245479</a>
rdar://problem/100252023

Reviewed by Chris Dumez.

StorageAreaMap should sync with backend again instead of clearing map when an error occurs in backend. Also, all
synchronization should not remove pending changes from map, because backend will handle the requests and StorageAreaMap
will receive reply for pending requests after that. Until StorageAreaMap receives error in reply, it should assume
the request succeeds and report value in pending requests to client.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::setItem):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.cpp:
(WebKit::StorageAreaMap::setItem):
(WebKit::StorageAreaMap::clear):
(WebKit::StorageAreaMap::ensureMap):
(WebKit::StorageAreaMap::didSetItem):
(WebKit::StorageAreaMap::applyChange):
(WebKit::StorageAreaMap::clearCache):
(WebKit::StorageAreaMap::disconnect):
(WebKit::StorageAreaMap::syncOneItem):
(WebKit::StorageAreaMap::syncItems):
(WebKit::StorageAreaMap::resetValues): Deleted.
(WebKit::StorageAreaMap::shouldApplyChangeForKey const): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.h:

Canonical link: <a href="https://commits.webkit.org/256863@main">https://commits.webkit.org/256863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b302b6213898b9d34f16b541e2659b99e40bf2dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97028 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30142 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106551 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6525 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35026 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89419 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103245 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102698 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83642 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/329 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/311 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4747 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5102 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1545 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40824 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -945,19 +945,22 @@ void NetworkStorageManager::cloneSessionStorageNamespace(IPC::Connection& connec
-        return completionHandler(hasQuotaError);
+        return completionHandler(hasError, WTFMove(allItems));
 
+    auto temp = key;
```
What is this for? It appears to be an unused variable.

szewai committed:
REGRESSION (iOS 16): LocalStorage is cleared after quota error
https://bugs.webkit.org/show_bug.cgi?id=245479
rdar://problem/100252023

Reviewed by Chris Dumez.

StorageAreaMap should sync with backend again instead of clearing map when an error occurs in backend. Also, all
synchronization should not remove pending changes from map, because backend will handle the requests and StorageAreaMap
will receive reply for pending requests after that. Until StorageAreaMap receives error in reply, it should assume
the request succeeds and report value in pending requests to client.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::setItem):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.cpp:
(WebKit::StorageAreaMap::setItem):
(WebKit::StorageAreaMap::clear):
(WebKit::StorageAreaMap::ensureMap):
(WebKit::StorageAreaMap::didSetItem):
(WebKit::StorageAreaMap::applyChange):
(WebKit::StorageAreaMap::clearCache):
(WebKit::StorageAreaMap::disconnect):
(WebKit::StorageAreaMap::syncOneItem):
(WebKit::StorageAreaMap::syncItems):
(WebKit::StorageAreaMap::resetValues): Deleted.
(WebKit::StorageAreaMap::shouldApplyChangeForKey const): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.h:

Canonical link: https://commits.webkit.org/256863@main
